server/routerlicious/packages/lambdas/src/deli/lambda.ts

```diff
@@ -245,14 +246,38 @@ export class DeliLambda implements IPartitionLambda {
         const message = rawMessage as IRawOperationMessage;
         const dataContent = this.extractDataContent(message);
 
-        // Check if we should nack all messages
-        if (this.nackFutureMessages) {
-            return this.createNackMessage(
-                message,
-                this.nackFutureMessages.code,
-                this.nackFutureMessages.type,
-                this.nackFutureMessages.message,
-                this.nackFutureMessages.retryAfter);
+        // Check if we should nack this message
+        const nackMessages = this.nackMessages;
+        if (nackMessages) {
+            let shouldNack = true;
+
+            if (nackMessages.allowSystemMessages && (isServiceMessageType(message.type) || !message.clientId)) {
+                // this is a system message. don't nack it
+                shouldNack = false;
+            } else if (nackMessages.allowedScopes) {
+                const clientId = message.clientId;
+                if (clientId) {
+                    const client = this.clientSeqManager.get(clientId);
+                    if (client) {
+                        for (const scope of nackMessages.allowedScopes) {
+                            if (client.scopes.includes(scope)) {
+                                // this client has an allowed scope. don't nack it
+                                shouldNack = false;
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
At least today, we send the summaryScope to everyone. Which means we are really not nacking anyone. I am wondering whether we should make it more granular; check the scope and also op type?

Reply:
It was hard to recover docs that ended up in a state where there were too many ops since the last summary. The solution we ended up with was to allow system ops for the main client and allow all ops for the summarizer client (by checking the client type is "summarizer"). Since that worked, I would rather only send the summaryScope for the summarizer client, that way deli will know which client is the summarizer.

Reply:
> I would rather only send the summaryScope for the summarizer client, that way deli will know which client is the summarizer.

I am not sure how this is doable. Today, every client needs the scope since they can get disconnected and we need to pick another client. And then that client needs to spawn a summarizer. As much as I know, it needs to inherit the scope from its parent so "parent scope" needs to be a superset of summarizers scope.

Checking "clientType" is an option but I am not sure how secure that is since its generated by the client.

@arinwt: Let me know if my understanding is correct

Reply:
I think everything you said makes sense. Checking the clientType prevents against accidental extraneous ops, but not intentional. Like Gary said, that is how it currently works as far as I know, and it is still valid/useful.

Reply:
@arinwt Can you confirm if scopes are involved if a client can be a leader & spawn a summarizer? Or maybe only the "read"/"write" connection mode is used? If we don't give the main client the `summary:write` scope, will they still be able to spawn a summarizer client?

Reply:
As far as I know the summary:write scope isn't used. We do verify that we have a write connection mode in the parent first though.

Reply:
It's essentially just checking if we are the oldest member in the Quorum, which only includes write clients.

Reply:
Thanks. So it sounds like only giving summary:write to the summarizer client connection will work.

Reply:
We can't do that from client side. A client with only "write" permission can't spawn another client with a more privileged permission. That defeats the purpose of having scope.

I am also curious to understand how Push generates scope. When do you generate scope? And how do you decide what to give to whom.

Reply:
We give `ScopeType.DocRead` to read connections and `ScopeType.DocRead, ScopeType.DocWrite, ScopeType.SummaryWrite` scopes to write connections.

After this change I can give `ScopeType.DocRead, ScopeType.DocWrite` to write connections, and `ScopeType.DocRead, ScopeType.DocWrite, ScopeType.SummaryWrite` to write connections by a summary client.

Is there a reason that would not work?

Reply:
So when a client sends a connectionMode = write, you give them write scope. And then the plan is to check clientType in your alfred and only give summary scope to that specific clientType?

Reply:
Yea

Reply:
That defeats the purpose of scope since it should really come from a token claim. Client should have no control over it. Cause nothing is preventing a random client to ask for summary scope. But it's also what Push has today so the change is not adding any additional security vulnerability.

I am fine with checking this one in but it does not help r11s/FRS. We don't decide based on what client send us, we only verify. For example, if we see a connectionMode = write, we check whether the token has write as well. I think eventually you want a similar model as well, where SPO will tell us what scopes are for each client and Push will only store/verify it.

FYI @curtisman

Reply:
@GaryWilber and I synced up and Push gets the permission from Sharepoint so its okay to trust the client since Push has authorized the client anyway.